mqtt/mqtt-broker/src/ready.rs

```diff
@@ -0,0 +1,180 @@
+use std::{collections::HashSet, fmt::Display, hash::Hash};
+
+use tokio::sync::broadcast::{self, Receiver, RecvError, Sender};
+use tracing::{debug, warn};
+
+/// `BrokerReady` component acts as a intermidiate entity to determine whether
+/// `Broker` is ready to serve external clients. It awaits on several events
+/// from components Broker depends on.
+///
+/// Notes: Authenticator and Authorizer both needs some data which not
+/// available when process started. They receive it using internal
+/// communication channels. Once data is loaded and components initialized
+/// they send corresponding events to the `BrokerReady` which in turn unblocks
+/// external transports.
+pub struct BrokerReady<E> {
+    event_send: Sender<E>,
+}
+
+impl<E> Default for BrokerReady<E> {
+    fn default() -> Self {
+        let (event_send, _) = broadcast::channel(5);
+        Self { event_send }
+    }
+}
+
+impl<E> BrokerReady<E> {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn handle(&self) -> BrokerReadyHandle<E> {
+        BrokerReadyHandle(self.event_send.clone())
+    }
+
+    pub fn signal(&self) -> BrokerReadySignal<E> {
+        BrokerReadySignal(self.event_send.subscribe())
+    }
+}
+
+/// A handle to send readiness signals to `BrokerReady`.
+pub struct BrokerReadyHandle<E>(Sender<E>);
+
+impl<E> BrokerReadyHandle<E>
+where
+    E: Clone + Display,
+{
+    pub fn send(&mut self, event: E) {
+        loop {
```

Comment:
should there be a wait time between sends?

Reply:
I made it a second thought it probably doesn't need a loop